### PR TITLE
Server-side Lua capability

### PR DESCRIPTION
**Pull request by blexaer**
Added capability to run server side lua for later use in API.

## How Has This Been Tested?
Loading of a test script with function that run on player connect/disconnect.

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: common/lua-system.cpp [hunk 1]
```diff
@@ -17,18 +18,24 @@ namespace tec {
...
 		for (auto itr = LuaScriptMap::Begin(); itr != LuaScriptMap::End(); itr++) {
-			auto entity_id = itr->first;
+			//auto entity_id = itr->first; <------ commented as it is currently unused
 			if (!itr->second->script_name.empty() && itr->second->environment["onUpdate"].valid()) {
 				itr->second->environment["onUpdate"](delta);
 			}
 		}
 	}
 
+	void LuaSystem::ProcessEvents() {
```
Comment (adam4813) [inline common/lua-system.cpp:31]: Nice cleanup!

File: common/lua-system.cpp [hunk 2]
```diff
@@ -62,10 +69,36 @@ namespace tec {
...
 
+	void LuaSystem::On(std::shared_ptr<ClientConnectedEvent> data) {
+		CallFunction("onClientConnected");
+	}
+	
+	void LuaSystem::On(std::shared_ptr<ClientDisconnectedEvent> data) {
+		CallFunction("onClientDisconnected");
+	}
+
+	void LuaSystem::CallFunction(std::string function_name) {
```
Comment (adam4813) [inline common/lua-system.cpp:80]: Good idea to make this interface with the script environment.

File: common/lua-system.cpp [hunk 3]
```diff
@@ -62,10 +69,36 @@ namespace tec {
...
+		// multiton <eid, LuaScript*>
+		for (auto itr = LuaScriptMap::Begin(); itr != LuaScriptMap::End(); itr++) {
+			if (!itr->second->script_name.empty() && itr->second->environment[function_name].valid()) {
+				itr->second->environment[function_name]();
+			}
+		}
+		// list<scripts>
+		for (auto script = scripts.begin(); script != scripts.end(); script++) {
```
Comment (adam4813) [inline common/lua-system.cpp:88]: So these scripts are sort of the "global" scripts and aren't bound to an entity?

File: common/lua-system.cpp [hunk 4]
```diff
@@ -62,10 +69,36 @@ namespace tec {
...
+		}
+	}
+
 	std::shared_ptr<LuaScript> LuaSystem::LoadFile(FilePath filepath) {
 		std::shared_ptr<LuaScript> script = std::make_shared<LuaScript>(ScriptFile::Create(filepath));
 		script->SetupEnvironment(&this->lua);
 		script->ReloadScript();
+
+		// add to list of scripts
+		scripts.push_back(*script);
```
Comment (adam4813) [inline common/lua-system.cpp:101]: Be cautious here as you are changing the behavior now. I see what your doing, and it is probably the better choice to do what you are doing.

**Review by adam4813 — COMMENTED** [review]
Everything I see here looks good so far and is a great starting point! We could almost merge this in already as it seems to stand alone (minus the events it is intended to actually respond to).

File: common/lua-system.cpp (hunk 3 above)
Comment (blexaer) [reply common/lua-system.cpp:88]: yes, at the moment

File: common/lua-system.cpp (hunk 1 above)
Comment (blexaer) [reply common/lua-system.cpp:31]: :-)

File: common/lua-system.cpp (hunk 4 above)
Comment (blexaer) [reply common/lua-system.cpp:101]: It could be stored as a single script instead where by this is avoided. Its comes down to how the API should be setup.

**Review by adam4813 — APPROVED** [review]

**Review by adam4813 — APPROVED** [review]

File: common/lua-system.cpp [hunk 5]
```diff
@@ -37,7 +42,8 @@ void LuaSystem::On(std::shared_ptr<EntityCreated> data) {
 			script->SetupEnvironment(&this->lua);
 			script->In(comp);
 			LuaScriptMap::Set(entity_id, script);
-		} break;
+		}
+		break;
```
Comment (adam4813) [inline common/lua-system.cpp:46]: Can you put this inside the curly just above it?

File: common/lua-system.hpp [hunk 6]
```diff
@@ -13,29 +13,32 @@
 #include "tec-types.hpp"
 
 namespace tec {
-class LuaSystem;
-typedef Command<LuaSystem> LuaCommand;
-struct EntityCreated;
-struct EntityDestroyed;
+	class LuaSystem;
```
Comment (adam4813) [inline common/lua-system.hpp:16]: Got some extra indent inside the namespace in this file

File: server/main.cpp [hunk 7]
```diff
@@ -41,16 +44,28 @@ std::string LoadJSON(const FilePath& fname) {
 	return in;
 }
 
+// Loads a given json file into the system
 void ProtoLoadEntity(const FilePath& fname) {
-	std::shared_ptr<EntityCreated> data = std::make_shared<EntityCreated>();
```
Comment (adam4813) [inline server/main.cpp:45]: Space instead of tabs in this function

File: common/events.hpp [hunk 8]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <commands.pb.h>
+#include <asio.hpp>
```
Comment (adam4813) [inline common/events.hpp:4]: Unused include?

Comment (blexaer) [reply common/events.hpp:4]: Gone

File: common/lua-system.cpp (hunk 5 above)
Comment (blexaer) [reply common/lua-system.cpp:46]: Yes

File: common/lua-system.hpp (hunk 6 above)
Comment (blexaer) [reply common/lua-system.hpp:16]: Done

File: server/main.cpp (hunk 7 above)
Comment (blexaer) [reply server/main.cpp:45]: Fixed

File: server/main.cpp [hunk 9]
```diff
@@ -50,7 +54,18 @@ void ProtoLoadEntity(const FilePath& fname) {
 }
 } // namespace tec
 
+std::shared_ptr<spdlog::logger> tec_log;
+void InitializeLogger() {
+    std::vector<spdlog::sink_ptr> sinks;
```
Comment (adam4813) [inline server/main.cpp:59]: Looks like this one may have gotten spaces instead of tabs as well

**Review by adam4813 — COMMENTED** [review]
Minor indent change, but looks good

File: common/lua-system.hpp [hunk 10]
```diff
@@ -34,8 +35,10 @@ class LuaSystem : public CommandQueue<LuaSystem>, public EventQueue<EntityCreate
...
 	sol::state& GetGlobalState() { return this->lua; }
-
 private:
 	sol::state lua;
+	std::list<LuaScript> scripts;
+
+	void CallFunction(std::string);
 };
-} // namespace tec
+} // namespace tec
```
Comment (Meisaka) [inline common/lua-system.hpp:44]: The newline went missing.

Comment (blexaer) [reply common/lua-system.hpp:44]: adding newline

Comment (blexaer) [reply common/lua-system.hpp]: I've put it back

**Review by Meisaka — APPROVED** [review]
All looks good to me.

**Pull request merged**